src/Pull.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+#include <functional>
+#include <memory>
+#include <nanomsg/pipeline.h>
+#include <nanomsg/nn.h>
+#include <string>
+#include <vector>
+#include "NanoProtocol.h"
+class Pull {
+public:
+   Pull(const std::string& location, const bool shouldConnect=false);
+   Pull(const std::string& location,
```

Comment:
I would like a discussion with you and @john-gress  (and anyone else who is interested). 
Not sure I'm a big fan of the logic in the constructors for ```Pull```, ```Push```.   I can see there are ```bind``` (connection owner) considerations but I would much rater extract this out to factory methods. Especially since they are used in smart pointers anyhow in Research. 

I won't comment for ```Push``` for the same thing but this comment goes for ```Push`` as well.